lib/fizzy/parser.hpp

```diff
@@ -21,125 +21,106 @@ using parser_result = std::tuple<T, const uint8_t*>;
 Module parse(bytes_view input);
 parser_result<Code> parse_expr(const uint8_t* input);
 
+template <typename T>
+parser_result<T> parse(const uint8_t* pos);
+
 template <typename T>
 struct parser
 {
 };
 
 template <>
-struct parser<uint8_t>
+inline parser_result<uint8_t> parse(const uint8_t* pos)
 {
-    parser_result<uint8_t> operator()(const uint8_t* pos)
-    {
-        const auto result = *pos;
-        ++pos;
-        return {result, pos};
-    }
-};
+    const auto result = *pos;
+    ++pos;
+    return {result, pos};
+}
 
 template <>
-struct parser<uint32_t>
+inline parser_result<uint32_t> parse(const uint8_t* pos)
 {
-    parser_result<uint32_t> operator()(const uint8_t* pos) { return leb128u_decode<uint32_t>(pos); }
-};
+    return leb128u_decode<uint32_t>(pos);
+}
 
 template <>
-struct parser<ValType>
+inline parser_result<ValType> parse(const uint8_t* pos)
 {
-    parser_result<ValType> operator()(const uint8_t* pos)
+    const auto b = *pos++;
+    switch (b)
     {
-        const auto b = *pos++;
-        switch (b)
-        {
-        case 0x7F:
-            return {ValType::i32, pos};
-        case 0x7E:
-            return {ValType::i64, pos};
-        default:
-            throw parser_error{"invalid valtype " + std::to_string(b)};
-        }
+    case 0x7F:
+        return {ValType::i32, pos};
+    case 0x7E:
+        return {ValType::i64, pos};
+    default:
+        throw parser_error{"invalid valtype " + std::to_string(b)};
     }
-};
+}
 
-template <>
-struct parser<Limits>
+inline parser_result<Limits> parse_limits(const uint8_t* pos)
```

Comment:
This is not a template specialization as it is never used in `parse_vec()`. But we can make it so if easier to read.